dds/idl/typeobject_generator.cpp

```diff
@@ -1180,250 +1720,598 @@ typeobject_generator::strong_connect(AST_Type* type, const std::string& anonymou
+      std::memcpy(complete_ti.sc_component_id().sc_component_id.hash, result, sizeof(OpenDDS::XTypes::EquivalenceHash));
 
       {
         size_t idx = 0;
-        for (List::const_iterator pos = scc.begin(), limit = scc.end(); pos != limit; ++pos) {
-          ti.sc_component_id().scc_index = static_cast<int>(++idx);
-          minimal_type_identifier_map_[pos->type] = ti;
-          minimal_type_map_[ti] = minimal_type_object_map_[pos->type];
+        for (List::const_iterator pos = scc.begin(); pos != scc.end(); ++pos) {
+          minimal_ti.sc_component_id().scc_index = static_cast<int>(++idx);
+          complete_ti.sc_component_id().scc_index = minimal_ti.sc_component_id().scc_index;
+          hash_type_identifier_map_[pos->type] = {minimal_ti, complete_ti};
         }
       }
+
+      // Compute the final type objects with the final type identifiers.
+      for (List::const_iterator pos = scc.begin(); pos != scc.end(); ++pos) {
+        generate_type_identifier(pos->type, true);
+        const OpenDDS::XTypes::TypeIdentifier& minimal_ti = hash_type_identifier_map_[pos->type].minimal;
+        const OpenDDS::XTypes::TypeIdentifier& complete_ti = hash_type_identifier_map_[pos->type].complete;
+        minimal_type_map_[minimal_ti] = type_object_map_[pos->type].minimal;
+        complete_type_map_[complete_ti] = type_object_map_[pos->type].complete;
+      }
     }
   }
 }
 
 void
-typeobject_generator::generate_minimal_type_identifier(AST_Type* type, bool force_type_object)
+typeobject_generator::update_maps(AST_Type* type,
+                                  const OpenDDS::XTypes::TypeObject& minimal_to,
+                                  const OpenDDS::XTypes::TypeObject& complete_to)
+{
+  const TypeObjectPair to_pair = {minimal_to, complete_to};
+  type_object_map_[type] = to_pair;
+
+  // In case of SCC, type identifiers of the types in SCC are computed first and
+  // type objects are constructed using them. In that case, we don't want to update
+  // each individual type's type identifier by hashing the corresponding type object.
+  // On the other hand, type identifier is computed if the type is not part of a SCC.
+  if (hash_type_identifier_map_.count(type) == 0) {
+    const OpenDDS::XTypes::TypeIdentifier minimal_ti = makeTypeIdentifier(minimal_to);
+    const OpenDDS::XTypes::TypeIdentifier complete_ti = makeTypeIdentifier(complete_to);
+    const TypeIdentifierPair ti_pair = {minimal_ti, complete_ti};
+    hash_type_identifier_map_[type] = ti_pair;
+
+    minimal_type_map_[minimal_ti] = minimal_to;
+    complete_type_map_[complete_ti] = complete_to;
+  }
+}
+
+void typeobject_generator::set_builtin_member_annotations(AST_Decl* member,
+  OpenDDS::XTypes::Optional<OpenDDS::XTypes::AppliedBuiltinMemberAnnotations>& annotations)
 {
-  // Generate the minimal identifier (and type object) and cache.
-  switch (type->node_type()) {
+  // Support only @hashid annotation for member at this time.
```

Comment:
Create JIRA ticket.

Reply:
We don't support the other "builtin" annotations (`@unit`, `@max`, `@min`, `@range`) at all, so I don't think we should add a ticket for this unless we also want make it part of a ticket for the other annotations in general. I'm also not sure why these are "builtin" since they have no bearing on the operation of DDS and aren't even mentioned anywhere else in the XTypes spec as far as I can tell. So the only spec that even defines them is the IDL spec.

Reply:
Like Fred said, this comment is to note that the IDL compiler currently doesn't support the other builtin annotations besides `@hashid` that are used in`AppliedBuiltinMemberAnnotations`. It isn't meant to say that this specific code doesn't support the other builtin annotations while they are already implemented.